### PR TITLE
Address Issue 21: Remove instrumentation boilerplate code

### LINKED ISSUE #21: Remove instrumentation code boilerplate
Most of the instrumentation code is generic across all snarks. E.g. the parameters for Fractal

```
            ("help", "print this help message")
            ("log_n_min", po::value<std::size_t>(&log_n_min)->default_value(8))
            ("log_n_max", po::value<std::size_t>(&log_n_max)->default_value(20))
            ("security_level", po::value<std::size_t>(&security_level)->default_value(128))
            ("field_size", po::value<std::size_t>(&field_size)->default_value(192))
            ("heuristic_ldt_reducer_soundness", po::value<bool>(&heuristic_ldt_reducer_soundness)->default_value(true))
            ("heuristic_fri_soundness", po::value<bool>(&heuristic_fri_soundness)->default_value(true))
            ("make_zk", po::value<bool>(&make_zk)->default_value(false))
            ("hash_enum", po::value<std::size_t>(&hash_enum_val)->default_value((size_t) libiop::blake2b_type))             /* Find a better solution for this in the future */
            ("is_multiplicative", po::value<bool>(&is_multiplicative)->default_value(false))
            ("optimize_localization", po::value<bool>(&optimize_localization)->default_value(false));
```

are all applicable to every SNARK. (And moreover, should really include `RS_extra_dimensions`.)

We should de-duplicate this, and in a future PR de-duplicate/improve the command line UX of this, and ease of instantiating SNARK parameters

### PR DESCRIPTION
**Pull request by pwang00**
Closes #21

### CODE REVIEW

File: libiop/profiling/boost_profile.cpp [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	    ("heuristic_fri_soundness", po::value<bool>(&options.heuristic_fri_soundness)->default_value(options.heuristic_fri_soundness)) /* Find a better solution for this in the future */
+	    ("is_multiplicative", po::value<bool>(&options.is_multiplicative)->default_value(options.is_multiplicative))
+		("make_zk", po::value<bool>(&options.make_zk)->default_value(false))
+		("optimize_localization", po::value<bool>(&options.optimize_localization)->default_value(false))
+		("hash_enum", po::value<std::size_t>(&options.hash_enum_val)->default_value((size_t) blake2b_type))
+  		("localization_parameter", po::value<std::size_t>(&options.localization_parameter)->default_value(2), "Only used when num_localization_steps is 0")
+        ("num_localization_steps", po::value<std::size_t>(&options.num_localization_steps)->default_value(0))
+        ("num_oracles", po::value<std::size_t>(&options.num_oracles)->default_value(1))
+        ("num_interactive_repetitions", po::value<std::size_t>(&options.num_interactive_repetitions)->default_value(1))
+        ("num_query_repetitions", po::value<std::size_t>(&options.num_query_repetitions)->default_value(64));
```
Comment (ValarDragon) [inline libiop/profiling/boost_profile.cpp:31]: Nice work on splitting this out!

However, this adds many new parameters to all the SNARKs, many of which don't belong or make sense.
Can this be changed to just be the intersection of what all SNARKs use, and then each SNARK still registers additional things they require?

Eventually I imagine we should have two such 'refactored program options', one for general things, and one for just FRI. (But for now lets just go with the intersection of what they all use)

Comment (pwang00) [reply libiop/profiling/boost_profile.cpp]: Ah I gotcha.  I think my latest commit fixes this, so whenever you have the chance feel free to give it a look!

File: libiop/profiling/boost_profile.cpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
+#ifndef CPPDEBUG /* Ubuntu's Boost does not provide binaries compatible with libstdc++'s debug mode so we just reduce functionality here */
+#include <boost/program_options.hpp>
+#endif
+
+#include "libiop/bcs/bcs_common.hpp"
+
+namespace po = boost::program_options;
+using namespace libiop;
+
+typedef struct options{
```
Comment (ValarDragon) [inline libiop/profiling/boost_profile.cpp:10]: Can you put a comment saying that its up to the caller to initialize `hash_enum` from `hash_enum_val`

File: libiop/profiling/boost_profile.cpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    std::size_t log_n_min = 8;
+    std::size_t log_n_max = 20;
+    std::size_t security_level = 128;
+    std::size_t field_size = 181;
+    std::size_t hash_enum_val = (size_t) libiop::blake2b_type;
+    bool heuristic_ldt_reducer_soundness = true;
+    bool is_multiplicative = true;
+    bool make_zk = false;
+    libiop::bcs_hash_type hash_enum = blake2b_type;
+} options;
```
Comment (ValarDragon) [inline libiop/profiling/boost_profile.cpp:20]: I don't think this trailing `options` is necessary. ref member initializers in: https://www.fluentcpp.com/2018/06/15/should-structs-have-constructors-in-cpp/

**Review by ValarDragon — COMMENTED** [review]
Looks great to me! Nice to see have the substantial code reduction =)

Happy to merge after the final two comments are addressed

File: libiop/profiling/boost_profile.cpp (hunk 3 above)
Comment (pwang00) [reply libiop/profiling/boost_profile.cpp]: Oops my bad, I think I used C struct syntax by accident. Should be fixed in the latest commit.

**Comment by ValarDragon** [comment]
Looks great to me! Thanks for the improvement =)

**Pull request merged**